This lab on Logistic Regression in R comes from p. 154-161 of "Introduction to Statistical Learning with Applications in R" by Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. It was re-implemented in Fall 2016 in `tidyverse` format by Amelia McNamara and R. Jordan Crouser at Smith College.

# 4.6.2 Logistic Regression

Let's return to the `Smarket` data from `ISLR`. 

In [1]:
list.of.packages <- c("car","ISLR","MASS")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

In [2]:
library(ISLR)
library(dplyr)
library(tidyr)
names(Smarket)
dim(Smarket)
summary(Smarket)


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



[1] "Year"      "Lag1"      "Lag2"      "Lag3"      "Lag4"      "Lag5"     
[7] "Volume"    "Today"     "Direction"

[1] 1250    9

      Year           Lag1                Lag2                Lag3          
 Min.   :2001   Min.   :-4.922000   Min.   :-4.922000   Min.   :-4.922000  
 1st Qu.:2002   1st Qu.:-0.639500   1st Qu.:-0.639500   1st Qu.:-0.640000  
 Median :2003   Median : 0.039000   Median : 0.039000   Median : 0.038500  
 Mean   :2003   Mean   : 0.003834   Mean   : 0.003919   Mean   : 0.001716  
 3rd Qu.:2004   3rd Qu.: 0.596750   3rd Qu.: 0.596750   3rd Qu.: 0.596750  
 Max.   :2005   Max.   : 5.733000   Max.   : 5.733000   Max.   : 5.733000  
      Lag4                Lag5              Volume           Today          
 Min.   :-4.922000   Min.   :-4.92200   Min.   :0.3561   Min.   :-4.922000  
 1st Qu.:-0.640000   1st Qu.:-0.64000   1st Qu.:1.2574   1st Qu.:-0.639500  
 Median : 0.038500   Median : 0.03850   Median :1.4229   Median : 0.038500  
 Mean   : 0.001636   Mean   : 0.00561   Mean   :1.4783   Mean   : 0.003138  
 3rd Qu.: 0.596750   3rd Qu.: 0.59700   3rd Qu.:1.6417   3rd Qu.: 0.596750  
 Max. 

In this lab, we will fit a logistic regression model in order to predict `Direction` using `Lag1` through `Lag5` and `Volume`. The `glm()` function fits **generalized linear models**, a class of models that includes logistic regression. 

The syntax of the `glm()` function is similar to that of `lm()`, except that we must pass in the argument `family=binomial` in order to tell `R` to run a logistic regression rather than some other type of generalized linear model.

In [3]:
glm_fit = glm(Direction~Lag1+Lag2+Lag3+Lag4+Lag5+Volume,
              data = Smarket,
              family = binomial)
summary(glm_fit)


Call:
glm(formula = Direction ~ Lag1 + Lag2 + Lag3 + Lag4 + Lag5 + 
    Volume, family = binomial, data = Smarket)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.446  -1.203   1.065   1.145   1.326  

Coefficients:
             Estimate Std. Error z value Pr(>|z|)
(Intercept) -0.126000   0.240736  -0.523    0.601
Lag1        -0.073074   0.050167  -1.457    0.145
Lag2        -0.042301   0.050086  -0.845    0.398
Lag3         0.011085   0.049939   0.222    0.824
Lag4         0.009359   0.049974   0.187    0.851
Lag5         0.010313   0.049511   0.208    0.835
Volume       0.135441   0.158360   0.855    0.392

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1731.2  on 1249  degrees of freedom
Residual deviance: 1727.6  on 1243  degrees of freedom
AIC: 1741.6

Number of Fisher Scoring iterations: 3


The smallest p-value here is associated with `Lag1`. The negative coefficient
for this predictor suggests that if the market had a positive return yesterday,
then it is less likely to go up today. However, at a value of 0.145, the p-value
is still relatively large, and so there is no clear evidence of a real association
between `Lag1` and `Direction`.

If you want, you can use helper methods like `coefficients()` and `residuals()` in order to access specific parts of this fitted model:

In [4]:
coefficients(glm_fit)
residuals(glm_fit)

(Intercept)         Lag1         Lag2         Lag3         Lag4         Lag5 
-0.126000257 -0.073073746 -0.042301344  0.011085108  0.009358938  0.010313068 
      Volume 
 0.135440659

1          2          3          4          5          6          7 
 1.1653998  1.2090622 -1.1455295  1.1516568  1.1591497  1.1656159 -1.1649514 
         8          9         10         11         12         13         14 
 1.1617721  1.1476292  1.1964318 -1.1715063 -1.2112126  1.1464685  1.1835565 
        15         16         17         18         19         20         21 
-1.1545426  1.1514147 -1.1865927  1.1236008 -1.2059449 -1.1745699 -1.2167138 
        22         23         24         25         26         27         28 
-1.2153152  1.1200356 -1.1640146 -1.1788969  1.1450922 -1.1637961 -1.2416508 
        29         30         31         32         33         34         35 
-1.2763075 -1.2768483 -1.2300622  1.1632691 -1.1291822 -1.1688465  1.0982354 
        36         37         38         39         40         41         42 
-1.2220802  1.1430560  1.1940514  1.2184076  1.2028873 -1.1739233 -1.2699626 
        43         44         45         46         47         48         49 
 0.9816748 -1.2261597  1.1266252 -1.2036326  1.1176085 -1.1682161 -1.2201056 
        50         51         52         53         54         55         56 
-1.3055894  1.1009430  1.2174395  1.2581652 -1.1008428 -1.2425614  1.0782339 
        57         58         59         60         61         62         63 
-1.1760375 -1.2149107 -1.3283866  1.0833752 -1.0584967  1.1950602  1.1548329 
        64         65         66         67         68         69         70 
-1.1104188  1.1837171 -1.1515868  1.1821467  1.1750792  1.2570978 -1.1115802 
        71         72         73         74         75         76         77 
-1.2262214 -1.2711210  1.0800524  1.1943109  1.2123759 -1.1281804  1.1728945 
        78         79         80         81         82         83         84 
 1.1840821 -1.1809429  1.1097105 -1.1725589 -1.1594646 -1.1931302 -1.2044441 
        85         86         87         88         89         90         91 
-1.1982039  1.1587257  1.1787624  1.1809151  1.2416677  1.2147997  1.1644370 
        92         93         94         95         96         97         98 
-1.1494345 -1.1903805  1.1038368 -1.2117226 -1.2012708 -1.2228242  1.1135058 
        99        100        101        102        103        104        105 
 1.1584995  1.2114814  1.2125069 -1.1363603  1.1538759 -1.1984884 -1.1874833 
       106        107        108        109        110        111        112 
 1.1342986 -1.1938561 -1.2131347 -1.2643083 -1.2558122  1.1560369  1.1786978 
       113        114        115        116        117        118        119 
 1.1982822 -1.1597198 -1.2046633 -1.2290985 -1.2115070  1.1483600 -1.1628885 
       120        121        122        123        124        125        126 
 1.1487267 -1.1635673 -1.1496349 -1.2292525  1.0745513 -1.2028997 -1.2123457 
       127        128        129        130        131        132        133 
 1.1337604  1.2371045 -1.1237600  1.1443491 -1.1956308  1.1483369 -1.1955083 
       134        135        136        137        138        139        140 
-1.1876308 -1.2432972  1.0852662  1.1843554  1.2386515 -1.1501392  1.1754053 
       141        142        143        144        145        146        147 
 1.1690008  1.1708593 -1.1779603 -1.1907049  1.1338622 -1.1933237 -1.2312106 
       148        149        150        151        152        153        154 
 1.1415429  1.2051451 -1.1535122 -1.1846269  1.1375700 -1.1892915  1.1336239 
       155        156        157        158        159        160        161 
-1.1770905  1.1580011 -1.1864617  1.1879064 -1.1303236 -1.1483750 -1.2478828 
       162        163        164        165        166        167        168 
-1.2459325  1.0932501 -1.1815292 -1.1671822 -1.1981557 -1.2686091  1.0528372 
       169        170        171        172        173        174        175 
-1.2003574 -1.3869797 -1.3135493 -1.2789214 -1.3489047  1.0191792  1.2414227 
       176        177        178        179        180        181        182 
-1.0923760  1.1462727  1.161654

The `predict()` function can be used to predict the probability that the
market will go up, given values of the predictors. The `type="response"`
option tells `R` to output probabilities of the form `P(Y = 1|X)`, as opposed
to other information such as the `logit`. If no data set is supplied to the
`predict()` function, then the probabilities are computed for the training
data that was used to fit the logistic regression model. 

In [5]:
glm_probs = data.frame(probs = predict(glm_fit, type="response"))
head(glm_probs)

probs    
1 0.5070841
2 0.4814679
3 0.4811388
4 0.5152224
5 0.5107812
6 0.5069565

Here we have printed
only the first ten probabilities. We know that these values correspond to
the probability of the market going up, rather than down, because the
`contrasts()` function indicates that `R` has created a dummy variable with
a 1 for `Up`.

In [6]:
contrasts(Smarket$Direction)
?contrasts

Up
Down 0 
Up   1

In order to make a prediction as to whether the market will go up or
down on a particular day, we must convert these predicted probabilities
into class labels, `Up` or `Down`. The following two commands create a vector
of class predictions based on whether the predicted probability of a market
increase is greater than or less than 0.5.

In [7]:
glm_pred = glm_probs %>%
  mutate(pred = ifelse(probs>.5, "Up", "Down"))
glm_probs %>%
  mutate(pred = ifelse(probs>.5, "Up", "Down"))

probs     pred
1    0.5070841 Up  
2    0.4814679 Down
3    0.4811388 Down
4    0.5152224 Up  
5    0.5107812 Up  
6    0.5069565 Up  
7    0.4926509 Down
8    0.5092292 Up  
9    0.5176135 Up  
10   0.4888378 Down
11   0.4965211 Down
12   0.5197834 Up  
13   0.5183031 Up  
14   0.4963852 Down
15   0.4864892 Down
16   0.5153660 Up  
17   0.5053976 Up  
18   0.5319322 Up  
19   0.5167163 Up  
20   0.4983272 Down
21   0.5229797 Up  
22   0.5221677 Up  
23   0.5340639 Up  
24   0.4920971 Down
25   0.5008751 Up  
26   0.5191211 Up  
27   0.4919679 Down
28   0.5373796 Up  
29   0.5571306 Up  
30   0.5574363 Up  
⋮    ⋮         ⋮   
1221 0.5437027 Up  
1222 0.5267398 Up  
1223 0.5329257 Up  
1224 0.5279180 Up  
1225 0.5350592 Up  
1226 0.5288983 Up  
1227 0.4893357 Down
1228 0.5534329 Up  
1229 0.5567250 Up  
1230 0.5592505 Up  
1231 0.5400175 Up  
1232 0.5230599 Up  
1233 0.5529143 Up  
1234 0.5413269 Up  
1235 0.5495887 Up  
1236 0.5495068 Up  
1237 0.5271413 Up  
1238 0.5263293 Up  
1239 0.5375299 Up  
1240 0.5281697 Up  
1241 0.5428337 Up  
1242 0.5650049 Up  
1243 0.5587216 Up  
1244 0.5425798 Up  
1245 0.5313870 Up  
1246 0.5198924 Up  
1247 0.5059256 Up  
1248 0.5392683 Up  
1249 0.5261183 Up  
1250 0.5179166 Up

The first command creates a vector of 1,250 `Down` elements. The second line
transforms to `Up` all of the elements for which the predicted probability of a
market increase exceeds 0.5. Given these predictions, we can `count()` how many observations were correctly or incorrectly classified.

In [8]:
glm_pred = cbind(Smarket, glm_pred)

glm_pred %>% 
  count(pred, Direction) %>%
  spread(Direction, n, fill = 0)

glm_pred %>%
  summarize(score = mean(pred == Direction))
head(glm_pred)

pred Down Up 
1 Down 145  141
2 Up   457  507

score 
1 0.5216

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume Today  Direction probs    
1 2001  0.381 -0.192 -2.624 -1.055  5.010 1.1913  0.959 Up        0.5070841
2 2001  0.959  0.381 -0.192 -2.624 -1.055 1.2965  1.032 Up        0.4814679
3 2001  1.032  0.959  0.381 -0.192 -2.624 1.4112 -0.623 Down      0.4811388
4 2001 -0.623  1.032  0.959  0.381 -0.192 1.2760  0.614 Up        0.5152224
5 2001  0.614 -0.623  1.032  0.959  0.381 1.2057  0.213 Up        0.5107812
6 2001  0.213  0.614 -0.623  1.032  0.959 1.3491  1.392 Up        0.5069565
  pred
1 Up  
2 Down
3 Down
4 Up  
5 Up  
6 Up

The diagonal elements of the confusion matrix indicate correct predictions,
while the off-diagonals represent incorrect predictions. Hence our model
correctly predicted that the market would go up on 507 days and that
it would go down on 145 days, for a total of 507 + 145 = 652 correct
predictions. The `mean()` function can be used to compute the fraction of
days for which the prediction was correct. In this case, logistic regression
correctly predicted the movement of the market 52.2% of the time.

At first glance, it appears that the logistic regression model is working
a little better than random guessing. But remember, this result is misleading
because we trained and tested the model on the same set of 1,250 observations.
In other words, 100− 52.2 = 47.8% is the **training error rate**. As we
have seen previously, the training error rate is often overly optimistic — it
tends to underestimate the _test_ error rate. 

In order to better assess the accuracy
of the logistic regression model in this setting, we can fit the model
using part of the data, and then examine how well it predicts the held out
data. This will yield a more realistic error rate, in the sense that in practice
we will be interested in our model’s performance not on the data that
we used to fit the model, but rather on days in the future for which the
market’s movements are unknown.

Like we did with KNN, we will first create a vector corresponding
to the observations from 2001 through 2004. We will then use this vector
to create a held out data set of observations from 2005.

In [9]:
train = Smarket %>%
  filter(Year < 2005)

test = Smarket %>%
  filter(Year >= 2005)

We now fit a logistic regression model using only the subset of the observations
that correspond to dates before 2005, using the subset argument.
We then obtain predicted probabilities of the stock market going up for
each of the days in our test set—that is, for the days in 2005.

In [10]:
glm_fit = glm(Direction~Lag1+Lag2+Lag3+Lag4+Lag5+Volume, 
              data = train, 
              family = binomial)

glm_probs = data.frame(probs = predict(glm_fit, 
                                       newdata = test, 
                                       type="response"))

Notice that we have trained and tested our model on two completely separate
data sets: training was performed using only the dates before 2005,
and testing was performed using only the dates in 2005. Finally, we compute
the predictions for 2005 and compare them to the actual movements
of the market over that time period.

In [15]:
glm_pred = glm_probs %>%
  mutate(pred = ifelse(probs>.5, "Up", "Down"))

glm_pred = cbind(test, glm_pred)

glm_pred %>% 
  count(pred, Direction) %>%
  spread(Direction, n, fill = 0)

glm_pred %>%
  summarize(score = mean(pred == Direction),
            recip = mean(pred != Direction))
n


glm_pred %>% 
  count(pred, Direction) %>%
  spread(pred, n, fill = 0)


pred Down Up 
1 Down 35    35
2 Up   76   106

score     recip    
1 0.5595238 0.4404762

[1] 1000

Direction Down Up 
1 Down      35    76
2 Up        35   106

The $!=$ notation means **not equal to**, and so the last command computes
the test set error rate. The results are rather disappointing: the test error
rate is 52%, which is worse than random guessing! Of course this result
is not all that surprising, given that one would not generally expect to be
able to use previous days’ returns to predict future market performance.
(After all, if it were possible to do so, then the authors of this book [along with your professor] would probably
be out striking it rich rather than teaching statistics.)

We recall that the logistic regression model had very underwhelming pvalues
associated with all of the predictors, and that the smallest p-value,
though not very small, corresponded to `Lag1`. Perhaps by removing the
variables that appear not to be helpful in predicting `Direction`, we can
obtain a more effective model. After all, using predictors that have no
relationship with the response tends to cause a deterioration in the test
error rate (since such predictors cause an increase in variance without a
corresponding decrease in bias), and so removing such predictors may in
turn yield an improvement. 

In the space below, refit a logistic regression using just `Lag1` and `Lag2`, which seemed to have the highest predictive power in the original logistic regression model.

In [12]:
glm_fit_lag1_lag2 = glm(Direction~Lag1+Lag2, 
              data = train, 
              family = binomial)
 # Write your code to fit the new model here


# -----------------------------------
# This will test your new model; you
# don't need to change anything below
# -----------------------------------
glm_probs = data.frame(probs = predict(glm_fit_lag1_lag2, 
                                       newdata = test, 
                                       type = "response"))

glm_pred = glm_probs %>%
  mutate(pred = ifelse(probs>.5, "Up", "Down"))

glm_pred = cbind(test, glm_pred)

glm_pred %>% 
  count(pred, Direction) %>%
  spread(Direction, n, fill = 0)

glm_pred %>%
  summarize(score = mean(pred == Direction))

pred Down Up 
1 Down 35    35
2 Up   76   106

score    
1 0.5595238

Now the results appear to be more promising: 56% of the daily movements
have been correctly predicted. The confusion matrix suggests that on days
when logistic regression predicts that the market will decline, it is only
correct 50% of the time. However, on days when it predicts an increase in
the market, it has a 58% accuracy rate.

Finally, suppose that we want to predict the returns associated with **particular
values** of `Lag1` and `Lag2`. In particular, we want to predict Direction on a
day when `Lag1` and `Lag2` equal 1.2 and 1.1, respectively, and on a day when
they equal 1.5 and −0.8. We do this using the `predict()` function.

In [13]:
predict(glm_fit_lag1_lag2, 
        newdata = data.frame(Lag1 = c(1.2, 1.5), 
                             Lag2 = c(1.1, -0.8)), 
        type = "response")

1         2 
0.4791462 0.4960939

To get credit for this lab, play around with a few other values for `Lag1` and `Lag2`, and then post to [#lab4](https://sds293.slack.com/messages/C79BELETC) about what you found. If you're feeling adventurous, try fitting models with other subsets of variables to see if you can find a better one!